srml/support/procedural/src/storage/genesis_config/builder_def.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Builder logic definition used to build genesis storage.
+
+use srml_support_procedural_tools::syn_ext as ext;
+use proc_macro2::TokenStream;
+use syn::spanned::Spanned;
+use quote::{quote, quote_spanned};
+use super::super::{DeclStorageDefExt, StorageLineTypeDef};
+
+pub struct BuilderDef {
+	/// Contains:
+	/// * build block for storage with build attribute.
+	/// * build block for storage with config attribute and no build attribute.
+	/// * build block for extra genesis build expression.
+	pub blocks: Vec<TokenStream>,
+	/// The build blocks requires generic traits.
+	pub is_generic: bool,
+}
+
+impl BuilderDef {
+	pub fn from_def(scrate: &TokenStream, def: &DeclStorageDefExt) -> Self {
```

Comment:
Most of the code in this function is still duplicated. Mostly it is just the code for the providing the data that should be inserted, that is different.

Reply:
oh this was because builder data is `Vec<key, value>` whereas config data is `Vec<&key, &value>` but I can come with some factorisation I think

Reply:
improved, and also now builder output must be a vec for maps not any iterator ? I feel that it improves error message like:
```
error[E0308]: mismatched types
  --> srml/support/test/tests/genesisconfig.rs:28:36
   |
28 |         pub AppendableDM config(t) build(|config| { ()}): double_map u32, blake2_256(T::BlockNumber) => Vec<u32>;
   |                                          ^ expected struct `std::vec::Vec`, found ()
   |
   = note: expected type `&std::vec::Vec<(u32, <T as Trait>::BlockNumber, std::vec::Vec<u32>)>`
              found type `&()`
```

And if really one wants to make specific insert he can just use extra genesis.